include/boost/math/optimization/cma_es.hpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright Nick Thompson, 2024
+ * Use, modification and distribution are subject to the
+ * Boost Software License, Version 1.0. (See accompanying file
+ * LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+ */
+#ifndef BOOST_MATH_OPTIMIZATION_CMA_ES_HPP
+#define BOOST_MATH_OPTIMIZATION_CMA_ES_HPP
+#include <algorithm>
+#include <atomic>
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <random>
+#include <sstream>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <boost/math/optimization/detail/common.hpp>
+#include <boost/math/tools/assert.hpp>
+#if __has_include(<Eigen/Dense>)
+#include <Eigen/Dense>
+#else
+#error "CMA-ES requires Eigen."
+#endif
+
+// Follows the notation in:
+// https://arxiv.org/pdf/1604.00772.pdf
+// This is a (hopefully) faithful reproduction of the pseudocode in the arxiv review
+// by Nikolaus Hansen.
+// Comments referring to equations all refer to this arxiv review.
+// A slide deck by the same author is given here:
+// http://www.cmap.polytechnique.fr/~nikolaus.hansen/CmaTutorialGecco2023-no-audio.pdf
+// which is also a very useful reference.
+
+#ifndef BOOST_MATH_DEBUG_CMA_ES
+#define BOOST_MATH_DEBUG_CMA_ES 0
+#endif
+
+namespace boost::math::optimization {
+
+template <typename ArgumentContainer> struct cma_es_parameters {
+  using Real = typename ArgumentContainer::value_type;
+  ArgumentContainer lower_bounds;
+  ArgumentContainer upper_bounds;
+  size_t max_generations = 1000;
+  ArgumentContainer const *initial_guess = nullptr;
+  // In the reference, population size = \lambda.
+  // If the population size is zero, it is set to equation (48) of the reference
+  // and rounded up to the nearest multiple of threads:
+  size_t population_size = 0;
+  // In the reference, learning_rate = c_m:
+  Real learning_rate = 1;
+};
+
+template <typename ArgumentContainer>
+void validate_cma_es_parameters(cma_es_parameters<ArgumentContainer> &params) {
+  using Real = typename ArgumentContainer::value_type;
+  using std::isfinite;
+  using std::isnan;
+  using std::log;
+  using std::ceil;
+  using std::floor;
+
+  std::ostringstream oss;
+  detail::validate_bounds(params.lower_bounds, params.upper_bounds);
+  if (params.initial_guess) {
+    detail::validate_initial_guess(*params.initial_guess, params.lower_bounds, params.upper_bounds);
+  }
+  const size_t n = params.upper_bounds.size();
+  // Equation 48 of the arxiv review:
+  if (params.population_size == 0) {
+    //auto tmp = 4.0 + floor(3*log(n));
+    // But round to the nearest multiple of the thread count:
+    //auto k = static_cast<size_t>(std::ceil(tmp/params.threads));
+    //params.population_size = k*params.threads;
+    params.population_size = static_cast<size_t>(4.0 + floor(3*log(n)));
```

Comment:
```suggestion
    params.population_size = static_cast<size_t>(4 + floor(3*log(n)));
```

Some of the `<stdfloat>` types will warn/error on trying to add a double.